### DISCUSSION TITLE
Recommendations for porting existing hardware packages (e.g. Arduino, CircuitPython, MicroPython)

### DISCUSSION THREAD

**Discussion by sgbaird** [MrYsLab/telemetrix-rpi-pico#17] [Ideas]
For example, I've been using the [AS7341 light sensor](https://www.adafruit.com/product/4698), and it has [an Arduino package](https://www.arduino.cc/reference/en/libraries/adafruit-as7341/), [a CircuitPython package](https://docs.circuitpython.org/projects/as7341/en/latest/), and [an unofficial MicroPython version](https://gitlab.com/robhamerling/micropython-as7341/-/tree/main/examples). Any recommendations for using sensors with existing Arduino/CP/MP packages, but with `telemetrix-rpi-pico` or the Pico W variant?

**Comment by MrYsLab** [comment]
@sgbaird An Arduino package version would need to be ported using the telemetrix-rpi-pico API. The ported code should be able to be used with the Pico W version of telemetrix as well, or any other version of telemetrix. If you need a general-purpose library, the entire Arduino library would need to be ported, or if you need a subset of functionality, only that would be ported. For telemetrix, i2c is handled as an application.
So that you can see a demonstration of each type of port, a partial port would look like what I did for the **[ADXL345](https://github.com/MrYsLab/telemetrix-rpi-pico/blob/master/examples/i2c_adxl345_accelerometer.py)**. For a full port, you can check out this [LCD i2c library](https://github.com/MrYsLab/telemetrix/blob/master/examples/Contributed/i2c_lcd_pcf8574/i2c_lcd_pcf8574.py) that was ported in its entirety. The library is then used in [this example](https://github.com/MrYsLab/telemetrix/blob/master/examples/Contributed/i2c_lcd_pcf8574/i2c_lcd_pcf8574_hello_world.py).
The porting process is usually not too tricky but can be very tedious. If the original library includes delays, those times may need to be adjusted.
